cranelift/codegen/meta/src/isa/x86/recipes.rs

```diff
@@ -1703,6 +1704,7 @@ pub(crate) fn define<'shared>(
                         sink.put4(offset as u32);
                     "#,
                 ),
+            "size_plus_maybe_sib_inreg1_plus_rex_prefix_for_inreg0_inreg1",
```

Comment:
I've felt for a while that there surely must be a better way than creating custom functions for every combination of `sib x offset x rex x inregs x outregs`: what if we generated this code instead by adding `check_if_needs_rex`, `check_if_needs_sib_or_offset`, etc., on the `EncodingRecipeBuilder`. Then during meta generation we could create code like that in `size_plus_maybe_sib_inreg1_plus_rex_prefix_for_inreg0_inreg1`.

@julian-seward1, what are your thoughts on this in relation to a new backend? I would imagine that the recipe part of all this is not necessary, but is the "generation of REX-inferring code" still something that would be useful?

Reply:
For the new x64 back end, REX generation is pretty much a non-problem.  There are two functions that emit all "standard form" instructions.  That is, those which have the structure OTHER_PREFIXES,REX,OPCODES,MODRM,SIB,IMM.  There is one for the reg-reg case, `emit_REX_OPCODES_MODRM_encG_encE` and one for the reg-mem case, `emit_REX_OPCODES_MODRM_SIB_IMM_encG_memE`.

https://github.com/cfallin/wasmtime/blob/arm64/cranelift/codegen/src/isa/x64/inst.rs#L1656

https://github.com/cfallin/wasmtime/blob/arm64/cranelift/codegen/src/isa/x64/inst.rs#L1787

For both functions, you pass the opcodes, the 'G' register field, and the 'E' field, which is either another register, or it's a memory address (an amode), and the functions will figure out the REX prefix on their own, and then skip it if it is redundant.  You can give flags for a bit of special-case handling -- "retain teh REX prefix even if it is redundant", "set REX.W to zero (to indicate a 32-bit operation)", but that's about it.

I'm sure this will cover the vast majority of instructions we need to generate, up to and including SSE4.2.  It won't do AVX or anything after that since it has no way to generate VEX or EVEX prefixes; but when the day comes, we can cook up similar routines.

Using the two routines, the top level emit-an-x64-instruction routine becomes simple; basically it's just a bunch of suitably parameterised calls to the above functions.

https://github.com/cfallin/wasmtime/blob/arm64/cranelift/codegen/src/isa/x64/inst.rs#L1787

What's implemented in the file is tested and believed to be correct.  Starting at around line 2500 there are no less than 400 encoding tests.

There is no attempt made here to support the 32-bit x86 architecture.  If indeed we do ever support that, I would argue in favour of treating it as a totally separate, unrelated architecture.  The alternative is to complicate and slow down the x64 code, which sounds like a poor tradeoff.  This probably sounds like a pretty radical proposal which runs against existing practice.  This separate-target strategy worked well for the Valgrind x86 back ends.  Not so well for the front ends, because there was a lot of duplication; but they are much bigger and more complex than the back ends.